File: 210-course-schedule-ii/course-schedule-ii.py
class Solution:
    # let's collect answer as args
    # let's return has_cycle
    # if you haven't visited yet, you can go through the vertex again and simply extends answer
    # you can still clean up ads once you know it's doesn't have cycle because before it cleans up,it collects result
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        self.answer = []
        self.learned = set()
        self.ad_list = self.init_ad_list(numCourses, prerequisites)
        for node in self.ad_list:
            if node not in self.learned:
                if self.dfs(node, set()):
                    return []
        
        return self.answer

    def dfs(self, node, visited):
        if node in visited:
            return True
        visited.add(node)
        for ad in self.ad_list[node]:
            if self.dfs(ad, visited):
                return True
        self.ad_list[node] = []
        if node not in self.learned:
            self.answer.append(node)
            self.learned.add(node)
        visited.remove(node)
        return False

    def init_ad_list(self, numCourses, edges):
        ad_list = {i: [] for i in range(numCourses)}
        for v, ad in edges:
            ad_list[v].append(ad)
        return ad_list
        

File: 210-course-schedule-ii/course-schedule-ii.java
class Solution {
    private Map<Integer, List<Integer>> adList;
    private List<Integer> answer;
    private Set<Integer> visited;
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        adList = initAdList(numCourses, prerequisites);
        visited = new HashSet<>();
        answer = new ArrayList<>();
        for (int i=0; i < numCourses; i++) {
            if (!visited.contains(i)) {
                if (findCycle(i, new HashSet<>())) {
                    return new int[0];
                }
            }
        }
        Collections.reverse(answer);
        return answer.stream().mapToInt(Integer::intValue).toArray();
    }

    public boolean findCycle(int curr, Set<Integer> path) {
        //System.out.println(curr);
        if (path.contains(curr)) {
            return true;
        }
        path.add(curr);
        for (int ad: adList.get(curr)) {
            if (findCycle(ad, path)) {
                return true;
            }
        }
        adList.put(curr, new ArrayList<>());
        if (!visited.contains(curr)) {
            answer.add(curr);
        }
        path.remove(curr);
        visited.add(curr);
        return false;

    }

    public Map<Integer, List<Integer>> initAdList(int n, int[][] edges) {
        Map<Integer, List<Integer>> adList = new HashMap<>();
        for (int i = 0; i < n; i++) {
            adList.put(i, new ArrayList<>());
        }
        for (int[] edge: edges) {
            adList.get(edge[1]).add(edge[0]);
        }
        return adList;
    }
}